torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -68,95 +69,78 @@ def foreach_all_gather(
     group_rank = group.rank()
     if use_uint8:
         total_padded_unsharded_numel = sum(
-            fsdp_param._padded_unsharded_size.numel()
-            # Multiply by the dtype size to convert to uint8 numel
-            * fsdp_param.unsharded_param_data_dtype.itemsize
-            for fsdp_param in fsdp_params
+            fsdp_param.padded_unsharded_bytes for fsdp_param in fsdp_params
         )
     else:
         total_padded_unsharded_numel = sum(
-            fsdp_param._padded_unsharded_size.numel() for fsdp_param in fsdp_params
+            fsdp_param.padded_unsharded_numel for fsdp_param in fsdp_params
         )
     total_sharded_numel = total_padded_unsharded_numel // world_size
     with torch.cuda.stream(all_gather_copy_in_stream):
         # - Copy in
-        flat_tensor = torch.empty(
+        all_gather_output = torch.empty(
             (total_padded_unsharded_numel,), dtype=dtype, device=device
         )
-        flat_tensor_shards = [
```

Comment:
Since the copy-out now operates on the 1D `all_gather_output`, we do not need to compute views for each shard. We reduce this loop over `world_size` to just 1 `narrow()` call for the calling rank.